File: solutions/chapter4/problem_4_5.cpp
#include <bits/stdc++.h>
using namespace std;

int K;

int f(string n) {
    if (n != "") {
        if (stoll(n) > K) return 0;
    }

    bool three = false;
    bool five = false;
    bool seven = false;

    for (int i = 0; i < n.size(); i++) {
        if (n[i] == '3') {
            three = true;
        } else if (n[i] == '5') {
            five = true;
        } else {
            seven = true;
        }
    }

    return (three && five && seven) + f('3'+n) + f('5'+n) + f('7'+n);
}

int main() {
    cin >> K;
    cout << f("") << endl;
}

File: solutions/chapter4/problem_4_5.py
K = 0

def f(n):
    if n != "":
        if int(n) > K:
            return 0
    
    three, five, seven = False, False, False

    for x in n:
        if x == '3':
            three = True
        elif x == '5':
            five = True
        else:
            seven = True
    
    return (three and five and seven) + f('3'+n) + f('5'+n) + f('7'+n)


def main():
    global K
    K = int(input())
    print(f(""))

if __name__ == '__main__':
    main()